### PR TITLE
Mapnik use JSON/XML string for configuration

### PR DESCRIPTION
**Pull request by aashish24**

### CODE REVIEW

**Comment by aashish24** [comment]
- [ ] Add test

**Comment by aashish24** [comment]
@alculquicondor still some work to do but have a look at it and let me know if for most part the changes looks good to you.

**Comment by alculquicondor** [comment]
I havent checked everything yet, but my first thought is that we want to provide backwards compatibility. In this sense, we shouldn't change the name of the property `mapfile` or support both the new one and the old one.

**Comment by aashish24** [comment]
@alculquicondor that is possible, let me see if there is less-code way to supporting it. I have an idea.

**Comment by aashish24** [comment]
@alculquicondor I made my attempt on making it backward compatible. Have a look. Also, is there a way to test the mapnik instance?

**Comment by alculquicondor** [comment]
Currently, there is no tests for mapnik. We might need to install extra packages in travis.

**Comment by alculquicondor** [comment]
A good test would be to just generate an specific tile.

**Comment by aashish24** [comment]
> A good test would be to just generate an specific tile.

Right and for now we can verify if the tile exists on the FS. Okay, I will look into it some more.

**Comment by alculquicondor** [comment]
One thought: defining Mapnik with json is not an standard way to do it AFAIK. The standard way is with either XML or with mapnik objects https://github.com/mapnik/mapnik/wiki/GettingStartedInPython#create-a-map

File: TileStache/Mapnik.py [hunk 1]
```diff
@@ -248,7 +256,7 @@ def __init__(self, layer, mapfile, fields=None, layers=None, layer_index=0, scal
     def prepareKeywordArgs(config_dict):
         """ Convert configured parameters to keyword args for __init__().
         """
-        kwargs = {'mapfile': config_dict['mapfile']}
+        kwargs = {'mapconfig': config_dict['mapconfig']}
```
Comment (alculquicondor) [inline TileStache/Mapnik.py]: backwards compatibility issue here too

File: TileStache/Mapnik.py [hunk 2]
```diff
@@ -74,19 +74,24 @@ class ImageProvider:
...
-        """ Initialize Mapnik provider with layer and mapfile.
+    def __init__(self, layer, mapconfig, fonts=None, scale_factor=None):
+        """ Initialize Mapnik provider with layer and mapconfig.
 
-            XML mapfile keyword arg comes from TileStache config,
+            XML mapconfig keyword arg comes from TileStache config,
             and is an absolute path by the time it gets here.
         """
-        maphref = urljoin(layer.config.dirpath, mapfile)
+        maphref = urljoin(layer.config.dirpath, mapconfig)
```
Comment (alculquicondor) [inline TileStache/Mapnik.py]: I don't like how a non-path is being passed to this function. We need to detect that it is not file differently somehow.

File: TileStache/Mapnik.py [hunk 3]
```diff
@@ -74,19 +74,24 @@ class ImageProvider:
...
         scheme, h, path, q, p, f = urlparse(maphref)
 
+        self.mapconfig_is_string = False
```
Comment (alculquicondor) [inline TileStache/Mapnik.py]: This variable name is not quite clear, because a path is an string too. Also, from what I understand, the need is to avoid using extra files to configure server, so Mapnik objects should be favored instead of XML strings.

File: TileStache/Mapnik.py (hunk 2 above)
Comment (aashish24) [inline TileStache/Mapnik.py]: > I don't like how a non-path is being passed to this function. We need to detect that it is not file differently somehow

any suggestions? My idea was to look at the URI which I would is sufficient. The other option would be to try to open the file and if it fails then assume it is a xml string. But then a invalid path would be treated as xml string too.

Comment (alculquicondor) [inline TileStache/Mapnik.py]: Not sure, but first, have you seen this?

```python
>>> urlparse(urljoin(None, '<Map></Map>'))
ParseResult(scheme='', netloc='', path='<Map></Map>', params='', query='', fragment='')
```

`scheme` is still `''`, so it will still be considered a path.

File: TileStache/Mapnik.py [hunk 4]
```diff
@@ -219,22 +227,22 @@ class GridProvider:
...
-        maphref = urljoin(layer.config.dirpath, mapfile)
+        maphref = urljoin(layer.config.dirpath, mapconfig)
         scheme, h, path, q, p, f = urlparse(maphref)
 
         if scheme in ('file', ''):
-            self.mapfile = path
+            self.mapconfig = path
         else:
-            self.mapfile = maphref
+            self.mapconfig = maphref
```
Comment (alculquicondor) [inline TileStache/Mapnik.py]: BTW, this is supposed to be given the same treatment as MapnikProvider, right?

File: TileStache/Mapnik.py (hunk 2 above)
Comment (alculquicondor) [inline TileStache/Mapnik.py]: You could try to actually call `mapnik.load_map_from_string` here if it is an string. This way, we will support passing Mapnik objects as part of the configuration dictionary.

**Comment by aashish24** [comment]
@alculquicondor this is ready for your review now.

File: TileStache/Goodies/Providers/Cascadenik.py [hunk 5]
```diff
@@ -31,19 +31,19 @@ class Provider (ImageProvider):
...
-        ImageProvider.__init__(self, layer, mapfile, fonts)
+        ImageProvider.__init__(self, layer, mapconfig, fonts)
 
     def renderArea(self, width, height, srs, xmin, ymin, xmax, ymax, zoom):
         """ Mostly hand off functionality to Mapnik.ImageProvider.renderArea()
         """
         if self.mapnik is None:
             self.mapnik = mapnik.Map(0, 0)
-            load_map(self.mapnik, str(self.mapfile), self.workdir, cache_dir=self.workdir)
+            load_map(self.mapnik, str(self.mapconfig), self.workdir, cache_dir=self.workdir)
```
Comment (alculquicondor) [inline TileStache/Goodies/Providers/Cascadenik.py]: We might want not to convert this to string. I'm still thinking of supporting Mapnik objects.

File: TileStache/Goodies/Providers/MapnikGrid.py [hunk 6]
```diff
@@ -125,17 +125,17 @@ def save(self, out, format):
...
-def get_mapnikMap(mapfile):
-    """ Get a new mapnik.Map instance for a mapfile
+def get_mapnikMap(mapconfig):
+    """ Get a new mapnik.Map instance for a mapconfig
     """
     mmap = mapnik.Map(0, 0)
 
-    if exists(mapfile):
-        mapnik.load_map(mmap, str(mapfile))
+    if exists(mapconfig):
```
Comment (alculquicondor) [inline TileStache/Goodies/Providers/MapnikGrid.py]: This will fail if you pass an XML string, wouldn't it? And then it will try to do the lines below.

File: TileStache/Mapnik.py (hunk 2 above)
Comment (alculquicondor) [inline TileStache/Mapnik.py]: Did you give a thought to this?

File: TileStache/Mapnik.py [hunk 7]
```diff
@@ -308,7 +330,7 @@ def renderArea(self, width, height, srs, xmin, ymin, xmax, ymax, zoom):
             finally:
                 global_mapnik_lock.release()
 
-        logging.debug('TileStache.Mapnik.GridProvider.renderArea() %dx%d at %d in %.3f from %s', width, height, self.scale, time() - start_time, self.mapfile)
+        logging.debug('TileStache.Mapnik.GridProvider.renderArea() %dx%d at %d in %.3f from %s', width, height, self.scale, time() - start_time, self.mapconfig)
```
Comment (alculquicondor) [inline TileStache/Mapnik.py]: I'm not sure if we should send the whole XML to the debug line

File: examples/geotiff/server.py [hunk 8]
```diff
@@ -23,14 +23,69 @@
...
+                   }
+                ]
+             }
+          }
+       }
+    }
+
+
+    mapnik_config = tostring(bf.etree(mapnik_config_json)[0])
+    print mapnik_config
```
Comment (alculquicondor) [inline examples/geotiff/server.py]: remove this line

File: tests/config_tests.py [hunk 9]
```diff
@@ -26,4 +28,48 @@ def test_config(self):
...
+
+        config_content = {
+            "cache": {
+                "name": "Test",
+                "path": "/tmp/stache",
+                "umask": "0000"
+            },
+            "layers": {
+                "geotiff": {
+                    "provider": {"name": "mapnik", "mapfile": mapnik_config},
```
Comment (alculquicondor) [inline tests/config_tests.py]: mapfile?

Comment (aashish24) [inline tests/config_tests.py]: to test backward compatibility

File: examples/geotiff/server.py (hunk 8 above)
Comment (aashish24) [inline examples/geotiff/server.py]: will take it out

File: TileStache/Mapnik.py [hunk 10]
```diff
@@ -74,22 +74,31 @@ class ImageProvider:
...
 
+        self.mapnik = None
+        self.layer = layer
+
         if scheme in ('file', ''):
-            self.mapfile = path
+            if os.path.isfile(path):
+                self.mapconfig = path
+            else:
+                self.mapnik = mapnik.Map(0, 0)
```
Comment (aashish24) [inline TileStache/Mapnik.py]: probably. But is there a harm in doing so?

Comment (aashish24) [inline TileStache/Mapnik.py]: doing this? I am not too familiar with the code base.

File: TileStache/Goodies/Providers/Cascadenik.py (hunk 5 above)
Comment (aashish24) [inline TileStache/Goodies/Providers/Cascadenik.py]: Can we do it in a separate PR? I am not too familiar with Mapnik objects

File: TileStache/Goodies/Providers/MapnikGrid.py (hunk 6 above)
Comment (aashish24) [inline TileStache/Goodies/Providers/MapnikGrid.py]: let me have a look

File: TileStache/Mapnik.py (hunk 10 above)
Comment (alculquicondor) [inline TileStache/Mapnik.py]: just to avoid extra processing, but I was most concerned with the line below.

File: tests/config_tests.py (hunk 9 above)
Comment (alculquicondor) [inline tests/config_tests.py]: I see, ok

File: TileStache/Goodies/Providers/MapnikGrid.py (hunk 6 above)
Comment (aashish24) [inline TileStache/Goodies/Providers/MapnikGrid.py]: @alculquicondor this method will not be called if we are passing the XML string because this method is called only when we have a path.

**Comment by aashish24** [comment]
@alculquicondor I made a minor change to ensure we can handle edge cases / unknown cases. Let me know what you think of it and then I can finalize my changes.

File: TileStache/Goodies/Providers/Cascadenik.py (hunk 5 above)
Comment (alculquicondor) [inline TileStache/Goodies/Providers/Cascadenik.py]: sure

**Comment by alculquicondor** [comment]
The most important thing that we can not leave is a test that uses a file instead of XML string

**Comment by aashish24** [comment]
> The most important thing that we can not leave is a test that uses a file instead of XML string

The test I added is XML string and we do have a test that reads the config from file. Are you saying that you want to add a test that uses the Mapnik XML config from the disk? If that's critical that's easy to add if it follows the same code path as the in memory xml string.

File: TileStache/Mapnik.py (hunk 7 above)
Comment (aashish24) [inline TileStache/Mapnik.py]: I am thinking that for debugging purposes, you do want to see the xml config. Or how else you would see it?

**Comment by aashish24** [comment]
@alculquicondor how about now, I added a test and handling failures better now.

**Comment by aashish24** [comment]
@alculquicondor feedback?

**Comment by alculquicondor** [comment]
sorry @aashish24, I have to find some free time :(

**Comment by aashish24** [comment]
@alculquicondor hoping you will find some time over the weekend :smile:

**Comment by aashish24** [comment]
@alculquicondor any feedback?

File: TileStache/Mapnik.py [hunk 11]
```diff
@@ -109,7 +124,10 @@ def __init__(self, layer, mapfile, fonts=None, scale_factor=None):
     def prepareKeywordArgs(config_dict):
         """ Convert configured parameters to keyword args for __init__().
         """
-        kwargs = {'mapfile': config_dict['mapfile']}
+        try:
```
Comment (alculquicondor) [inline TileStache/Mapnik.py]: This can be accomplished more easily like so:
`config_dict.get('mapconfig') or config_dict.get('mapfile')`

File: TileStache/Mapnik.py [hunk 12]
```diff
@@ -74,22 +74,37 @@ class ImageProvider:
...
-        maphref = urljoin(layer.config.dirpath, mapfile)
-        scheme, h, path, q, p, f = urlparse(maphref)
+        self.mapnik = None
+        self.layer = layer
 
-        if scheme in ('file', ''):
-            self.mapfile = path
-        else:
-            self.mapfile = maphref
+        try:
```
Comment (aashish24) [inline TileStache/Mapnik.py]: @alculquicondor the developers don't have to look at this part of code if eveything works but if we go with two separation options, we will still have to converge at some point. How about this, we merge this branch and I will make another pass on it since I would like to clean up few other thigns as well.

**Comment by aashish24** [comment]
> your code. I'm thinking that these can potentially decrease performance otherwise.

this is just for debugging so I would think that would be okay?

File: tests/config_tests.py [hunk 13]
```diff
@@ -1,6 +1,10 @@
 from unittest import TestCase
 from TileStache import Core, parseConfig
 
+from . import utils
+
+import os
```
Comment (aashish24) [inline tests/config_tests.py]: sure,

**Comment by aashish24** [comment]
@alculquicondor thoughts? How about we merge this branch and I will make another PR that will separate the config into mapfile, mapstring, and potentially mapobject.

File: TileStache/Mapnik.py (hunk 12 above)
Comment (alculquicondor) [inline TileStache/Mapnik.py]: No, because later we will break backwards compatibility. We need to do it the right way from the beginning. Sure, devs don't need to look this code, but if something fails, we can't tell the user if the XML was not well formatted or the path was wrong, because we don't have a way to properly differentiate between the two possible cases.

**Comment by alculquicondor** [comment]
Logs should be a single line to be useful.

**Comment by aashish24** [comment]
@alculquicondor I think we don't need this branch anymore, so I am inclined to close this PR and push another one that falls in inline with what we want.

**Comment by alculquicondor** [comment]
Sure, thanks
